---
title: "Process RNAseq cell lines from Cell Model Passports (CMP)"
output:
  html_document:
    df_print: paged
---

github repo: saezlab/process_rnaseq_cellines

Data downloaded from: https://cellmodelpassports.sanger.ac.uk/downloads

RNAseq data (version): rnaseq_2019-04-15_1133.csv.gz 
                       Read counts and FPKM-values provided as processed by iRAP pipeline
                       
Steps followed to process the RNAseq:

1. Loading data, matrix extraction and exploratory analysis of raw data

2. Removing samples with unknown tissue

3. Imputing 0 to NA values

4. Removing non-expressed genes with average CPM <=0

5. Normalization through TMM method

6. Voom transformation

7. Batch correction of voom transformed data using ComBat

8. Merging duplicates with the mean


```{r}
library(data.table)
library(dplyr)
library(stringr)
library(limma)
library(sva)
library(edgeR)

library(ggplot2)
library(ggforce)
library(ggfortify)

###### set paths and variables ######

wd = '~/Projects/process_rnaseq_cellines/CMP' #working directory
setwd(wd)

#metasample = '../../00-CMP/metadata/model_list_2019-04-04_0916.csv' #vector containing the path to the metainformation about the cell lines
metasample = '../../00-CMP/metadata/model_list_2019-06-21_1535.csv'
rnadata = '../../00-CMP/expression/rnaseq_2019-04-15_1133.csv'#vector containing the path to the RNAseq file of cell lines


###### Loading datasets ######
  
cmpRAW <- fread(rnadata, sep = ',')

metaSMP <- read.delim(metasample, sep = ',')
```

Matrix extraction

Create a matrix Gene x Sample. As the original dataset from Cell Model Pasports contains samples from 2 different sources (Sanger and Broad),
the samples are separarated based on the source, as they have different number of genes.

```{r}
models <- unique(cmpRAW$model_id) #unique samples

EXP_rawcounts=NULL
#pb <- txtProgressBar(min = 0, max = length(models), style = 3) # create progress bar
for (i in 1:length(models)) { #for each of the unique samples
#  setTxtProgressBar(pb, i) #update progress bar
  model <- models[i]
  df <-  cmpRAW[which(cmpRAW$model_id==model),c('gene_id','read_count', 'dataset_name')] #get the relevant information
  dset <- unique(df$dataset_name)
  for(institute in dset){
    dfsub <- df[which(df$dataset_name==institute),c('gene_id', 'read_count')]
    colnames(dfsub) <- c('gene_id', model)
    if(is.null(EXP_rawcounts[[institute]])){
      EXP_rawcounts[[institute]] <- dfsub
    }else{
      EXP_rawcounts[[institute]] <- merge(EXP_rawcounts[[institute]],dfsub, by = 'gene_id')
    }
  }
}
#close(pb)

rm(pb, df, dfsub, i, dset, institute, model)


# transforming data.table to data.frame 
names(EXP_rawcounts) <- gsub("Sanger RNASeq", 'sanger', names(EXP_rawcounts))
names(EXP_rawcounts) <- gsub("Broad \\(CCLE\\) RNASeq", 'broad', names(EXP_rawcounts))

EXP_rawcounts[['sanger']] <- setDF(EXP_rawcounts[['sanger']], rownames=EXP_rawcounts[['sanger']]$gene_id)
EXP_rawcounts[['sanger']] <- EXP_rawcounts[['sanger']][,-1]

EXP_rawcounts[['broad']] <- setDF(EXP_rawcounts[['broad']], rownames=EXP_rawcounts[['broad']]$gene_id)
EXP_rawcounts[['broad']] <- EXP_rawcounts[['broad']][,-1]

# saveRDS(EXP_rawcounts, file = '1907_release1904/rnaseq_2019-06-26_geneVSmodel_list.rds')
# EXP_rawcounts <- readRDS('1907_release1904/rnaseq_2019-06-26_geneVSmodel_list.rds')

print(lapply(EXP_rawcounts, dim))

#### Exploration of data

gg.venn <- data.frame(x = c(-0.866, 0.866),
                      y = c(0, 0),
                      labels = c('Sanger', 'Broad'))

gg.data.samples <- list(shared = intersect(colnames(EXP_rawcounts$sanger), colnames(EXP_rawcounts$broad)),
                        sanger = as.character(colnames(EXP_rawcounts$sanger) [ which( !colnames(EXP_rawcounts$sanger)%in%colnames(EXP_rawcounts$broad) ) ]),
                        broad = as.character(colnames(EXP_rawcounts$broad) [ which( !colnames(EXP_rawcounts$broad)%in%colnames(EXP_rawcounts$sanger) ) ]) )


gg.sample <- as.data.frame(matrix(data=NA, nrow=3, ncol = 4, dimnames = list(NULL, c('label', 'x', 'y', 'count'))), stringsAsFactors = F)
gg.sample$label <- c('shared', 'Sanger', 'Broad')
gg.sample$x <- c(0, -1.5, 1.5)
gg.sample$y <- c(0, 0, 0)
gg.sample$count <- c(length(gg.data.samples$shared), length(gg.data.samples$sanger), length(gg.data.samples$broad))


gg.data.genes <- list(shared = intersect(rownames(EXP_rawcounts$sanger), rownames(EXP_rawcounts$broad)),
                        sanger = as.character(rownames(EXP_rawcounts$sanger) [ which( !rownames(EXP_rawcounts$sanger)%in%rownames(EXP_rawcounts$broad) ) ]),
                        broad = as.character(rownames(EXP_rawcounts$broad) [ which( !rownames(EXP_rawcounts$broad)%in%rownames(EXP_rawcounts$sanger) ) ]) )


gg.genes <- as.data.frame(matrix(data=NA, nrow=3, ncol = 4, dimnames = list(NULL, c('label', 'x', 'y', 'count'))), stringsAsFactors = F)
gg.genes$label <- c('shared', 'Sanger', 'Broad')
gg.genes$x <- c(0, -1.5, 1.5)
gg.genes$y <- c(0, 0, 0)
gg.genes$count <- c(length(gg.data.genes$shared), length(gg.data.genes$sanger), length(gg.data.genes$broad))

ggplot(gg.venn, aes(x0 = x, y0 = y, r = 1.5, fill = labels)) +
  geom_circle(alpha = .3, size = 1, colour = 'grey') +
  coord_fixed() +
  labs(fill = NULL) +
  theme_void() + theme(legend.position = 'bottom') +
  annotate("text", x = gg.sample$x, y = gg.sample$y, label = gg.sample$count , size = 10) +
  ggtitle('Samples CMPs (2019-04-15_1133)') + 
  scale_fill_manual(values = c('#E69F00', '#009E73',  '#0072B2'))


ggplot(gg.venn, aes(x0 = x, y0 = y, r = 1.5, fill = labels)) +
  geom_circle(alpha = .3, size = 1, colour = 'grey') +
  coord_fixed() +
  labs(fill = NULL) +
  theme_void() + theme(legend.position = 'bottom') +
  annotate("text", x = gg.genes$x, y = gg.genes$y, label = gg.genes$count , size = 10) +
  ggtitle('Samples CMPs (2019-04-15_1133)') + 
  scale_fill_manual(values = c('#E69F00', '#009E73',  '#0072B2'))

```

The next steps of the pipleline can be run either by merging the shared samples, or by treating them separately until the normalization phase.
The raw counts distribution and the sequencing depth is different between both datasets, so they will be merge after the normalitation.

```{r}
#### Exploratory analysis of the raw counts distribution and the sequencing depth ####

# Raw counts distribution

par(mfrow = c(1,2))
hist(log2(as.numeric(unlist(EXP_rawcounts[['sanger']]))), breaks = 100, main = 'Cell lines raw counts Sanger', xlab = 'log2(raw counts)')
hist(log2(as.numeric(unlist(EXP_rawcounts[['broad']]))), breaks = 100, main = 'Cell lines raw counts Broad', xlab = 'log2(raw counts)')

# Sequencing Depth

par(mfrow = c(1,2))
barplot(colSums(EXP_rawcounts$sanger), main = 'Cell lines raw counts Sanger', xlab = 'Sum raw counts per Sample')
barplot(colSums(EXP_rawcounts$broad), main = 'Cell lines raw counts Broad', xlab = 'Sum raw counts per Sample')


```

2 and 3. Remove samples with unknown tissue and imput 0 to NA values. There are no samples with unknown tissue or with NA values.

```{r}

#message('Remove samples with unknown tissue\n')
tissue = unique(metaSMP$tissue)
samples = intersect(unique(c(colnames(EXP_rawcounts[['sanger']]), colnames(EXP_rawcounts[['broad']]))),  
                    metaSMP$model_id[ which(metaSMP$tissue %in% tissue) ])
EXP_rawcounts <- lapply(EXP_rawcounts, function(EXP){EXP[ , colnames(EXP) %in% samples ]})

#message('Imputing NA (ZEROmethod) \n')
#Therea are none for this dataset
exp_na <- lapply(EXP_rawcounts, function(EXP){# save gene:sample pairs that are NAs
                index <- which(is.na(EXP), arr.ind=T)
                paste(rownames(EXP)[index[,'row']],  colnames(EXP)[index[,'col']], sep=':')
          })
if ( !is.null(do.call(c,lapply(exp_na,dim))) ){
  EXP_rawcounts <- lapply(EXP_rawcounts, function(EXP){EXP[ which(is.na(EXP)) ] = 0})
}

cat('Total genes and cell lines: \n')
print(lapply(EXP_rawcounts, dim))

```

4. Removing non-expressed genes with average CPM <=0. The limma-voom method assumes that rows with zero or very low counts have been removed

```{r}

exp.zero <- lapply(EXP_rawcounts, function(EXP){
  index = aveLogCPM(EXP) <= 0
  rownames(EXP)[index]
})

# saveRDS(exp.zero, file = '1907_release1904/rnaseq_2019-06-26_expZero_list.rds')
# exp.zero <- readRDS(file = '1907_release1904/rnaseq_2019-06-26_expZero_list.rds')

EXP_rawcounts <- lapply(EXP_rawcounts, function(EXP){
  keep = aveLogCPM(EXP) > 0
  EXP [ keep, ]
})

cat('Total genes and cell lines: \n')
print(lapply(EXP_rawcounts, dim))

# saveRDS(EXP_rawcounts, file = '1907_release1904/rnaseq_2019-06-26_rmZeroGenes_list.rds')
# EXP_rawcounts <- readRDS(file = '1907_release1904/rnaseq_2019-06-26_rmZeroGenes_list.rds')
```

5. Normalization through TMM method. Although it is also possible to give a matrix of counts directly to voom without TMM normalization, limma package reommends it. "TMM" is the weighted trimmed mean of M-values (to the reference) proposed by Robinson and Oshlack (2010), where the weights are from the delta method on Binomial data. If refColumn is unspecified, the library whose upper quartile is closest to the mean upper quartile is used.

6. Voom transformation

```{r}
EXP_rawcounts = lapply(EXP_rawcounts, function(EXP){DGEList(counts=EXP,genes=rownames(EXP))})
EXP_normcounts = lapply(EXP_rawcounts, function(EXP){calcNormFactors(EXP, method = 'TMM')})
# saveRDS(EXP_normcounts, file = '1907_release1904/rnaseq_2019-06-26_tmm_norm_list.rds')
# EXP_normcounts <- readRDS(file = 'rnaseq_2019-06-26_tmm_norm_list.rds')

EXP_voom = lapply(EXP_normcounts, function(EXP){voom(EXP, plot = T)$E}) # $E extract the "data corrected for variance"

```

7. Batch correction of voom transformed data using ComBat. Using "tisue" and duplicates as covariates

```{r}

# join all samples in a matrix. Samples and Genes from the different sets are labeled with S or B respectively.
genes = Reduce(intersect, list(rownames(EXP_voom$sanger), rownames(EXP_voom$broad)))

tmp <- lapply(EXP_voom, function(EXP, genes){
  EXP <- as.data.frame(cbind(genes=rownames(EXP), EXP),stringsAsFactors=F)
  EXP[ genes, ]},genes)

cat('Total genes and cell lines: \n')
print(lapply(tmp, dim))

colnames(tmp$sanger) = paste0('S.',colnames(tmp$sanger))
colnames(tmp$broad) = paste0('B.',colnames(tmp$broad))

EXP <- merge(tmp$sanger, tmp$broad, by.x = 'S.genes', by.y = 'B.genes', row)
rownames(EXP) <- EXP$S.genes
EXP <- EXP[,2:ncol(EXP)]
EXP <- data.matrix(EXP, rownames.force = T)

cat('Total genes and cell lines:', dim(EXP), '\n')

## variables
catgroup <- as.data.frame(cbind(id=colnames(EXP),sample=gsub('[S/B]\\.','',colnames(EXP)),dataset=substr(colnames(EXP),1,1)),stringsAsFactors = F)
catgroup <- merge(catgroup, metaSMP[,c("model_id","tissue","tissue_status", "cancer_type")], by.x='sample', by.y='model_id')
catgroup <- catgroup[,-1]

## PCA

exp.pca <- prcomp(t(EXP))
str(exp.pca)

exp.pca.variances <- ((exp.pca$sdev^2) / (sum(exp.pca$sdev^2)))*100

#png('xplr/rnaseq_2019-06-26_PCA_barplot.png', height = 800, width = 900, res = 150)
barplot(exp.pca.variances, cex.names=1, xlab=paste("Principal component (PC), 1-", length(exp.pca$sdev)), 
        ylab="Proportion of variation (%)", main="Scree plot", ylim=c(0,20))
#dev.off()

#png('xplr/rnaseq_2019-06-26_PCA.png', height = 800, width = 900, res = 150)
qplot(exp.pca$x[,1], exp.pca$x[,2], data=catgroup,
      color= dataset, geom = c('point'), label = dataset,#,'text'
      xlab=paste("PC1, ", round(exp.pca.variances[1], 2), "%"), 
      ylab=paste("PC2, ", round(exp.pca.variances[2], 2), "%"))
#dev.off()

##

#message('Batch correction of voom transfromed data using ComBat\n')

batch = substr(colnames(EXP),1,1)
cov_tissue = as.character(metaSMP$tissue[ match( gsub('[S/B]\\.','',colnames(EXP)), metaSMP$model_id) ])
duplicates = gsub('[S/B]\\.','',colnames(EXP)[duplicated( gsub('[S/B]\\.','',colnames(EXP)) )])
cov_duplicates = sapply(duplicates, function(id) (metaSMP$model_id[ match(gsub('[S/B]\\.','',colnames(EXP)), metaSMP$model_id) ] == id) + 0)
cov_duplicates[ is.na(cov_duplicates) ] = 0
covariates = model.matrix(~cov_tissue+cov_duplicates)
EXP_corvoom = ComBat(EXP, batch = batch, mod = covariates, par.prior = T, prior.plots = T)
saveRDS(EXP_corvoom, file = '1907_release1904/rnaseq_2019-06-26_voom_batchcor.rds')
# EXP_corvoom <- readRDS(file = 'rnaseq_2019-06-26_voom_batchcor.rds')


```


8.Merging duplicates with the mean

```{r}

exp.pca <- prcomp(t(EXP_corvoom))
str(exp.pca)

exp.pca.variances <- ((exp.pca$sdev^2) / (sum(exp.pca$sdev^2)))*100

#png('xplr/rnaseq_2019-06-26_PCAcomBat_barplot.png', height = 800, width = 900, res = 150)
barplot(exp.pca.variances, cex.names=1, xlab=paste("Principal component (PC), 1-", length(exp.pca$sdev)), 
        ylab="Proportion of variation (%)", main="Scree plot", ylim=c(0,20))
#dev.off()

#png('xplr/rnaseq_2019-06-26_PCAcomBat.png', height = 800, width = 900, res = 150)
qplot(exp.pca$x[,1], exp.pca$x[,2], data=catgroup,
      color= dataset, geom = c('point'), label = dataset,#,'text'
      xlab=paste("PC1, ", round(exp.pca.variances[1], 2), "%"), 
      ylab=paste("PC2, ", round(exp.pca.variances[2], 2), "%"))
#dev.off()

##

message('Merging duplicates with the mean')
samples = unique( gsub('[S/B]\\.','',colnames(EXP)) )
genes = rownames(EXP_corvoom)
X = matrix(NA, nrow = length(genes), ncol = length(samples), dimnames = list(genes, samples))
for( s in samples){
  x = EXP_corvoom[ , grep(s, colnames(EXP_corvoom)) ]
  if ( is.matrix(x) )
    x = apply(x, 1, mean)
  X[, s] = x
}
EXPmerged = X
#saveRDS(EXPmerged, file = '1907_release1904/rnaseq_2019-06-26_voom_batchcor_merged.rds')

```
